### PR TITLE
Providers update202208

### LINKED ISSUE #8082: cpasbien provider update
### Before submitting your issue:

URL update and code from torrent9 to make the URL configurable

well, the table selection doesn't work, any idea ?
```
torrent_rows = html.find("table", {"class": "table-corps"})
```
=> _Exception while searching Cpasbien. Error: name 'torrent_table' is not defined_

Branch/Commit: 2022.8.15
OS: docker
Browser: ff
Code:
```
diff --git a/sickchill/oldbeard/providers/cpasbien.py b/sickchill/oldbeard/providers/cpasbien.py
index 99fd87dcc..b82543544 100644
--- a/sickchill/oldbeard/providers/cpasbien.py
+++ b/sickchill/oldbeard/providers/cpasbien.py
@@ -1,4 +1,5 @@
-import re
+
+import validators

 from sickchill import logger
 from sickchill.helper.common import convert_size, try_int
@@ -15,24 +16,85 @@ class Provider(TorrentProvider):
         self.public = True
         self.minseed = 0
         self.minleech = 0
-        self.url = "http://www.cpasbien.ac"
+        self._original_url = "https://www.cpasbien.ch"
+        self._custom_url = None
+        self._used_url = None
+        self._recheck_url = True

         self.proper_strings = ["PROPER", "REPACK"]
         self.cache = tvcache.TVCache(self)
         self.ability_status = self.PROVIDER_BACKLOG

+    def _retrieve_dllink_from_url(self, inner_url, _type="torrent"):
+        # for instance, no magnet
+        if _type != "torrent":
+            return ""
+        data = self.get_url(inner_url, returns="text")
+        # js
+        map = { "torrent": r"redirect", "magnet": r"redir" }
+        regex = r".*?function\s+" + map[_type] + r"\(\).+?= '([^']+)'"
+        # href
+        res = { "torrent": "", "magnet": "" }
+        with BS4Parser(data, "html5lib") as html:
+            # for manual testing:
+            # data = open('test.html', 'r').read()
+            # html = BeautifulSoup(data, "html5lib")
+            # try js
+            scripts = html.head.findAll("script")
+            if len(scripts):
+                script = scripts[-1].string
+                matches = re.match(regex, script, re.S)
+                if matches:
+                    return matches[1]
+            # try href
+            download_btns = html.findAll("div", {"class": "btn-download"})
+            for btn in download_btns:
+                link = btn.find("a")["href"]
+                if link.startswith("javascript"):
+                    return ""
+                if link.startswith("magnet"):
+                    res["magnet"] = link
+                else:
+                    res["torrent"] = link
+        return res[_type]
+
+    def _get_custom_url(self):
+        return self._custom_url
+
+    def _set_custom_url(self, url):
+        if self._custom_url != url:
+            self._custom_url = url
+            self._recheck_url = True
+
+    def _get_provider_url(self):
+        if self._recheck_url:
+            if self.custom_url:
+                if validators.url(self.custom_url):
+                    self._used_url = self.custom_url
+                else:
+                    logger.warning("Invalid custom url set, please check your settings")
+
+            self._used_url = self._original_url
+
+        return self._used_url
+
+    def _set_provider_url(self, url):
+        self._used_url = url
+
+    url = property(_get_provider_url, _set_provider_url)
+    custom_url = property(_get_custom_url, _set_custom_url)
+
     def search(self, search_strings, age=0, ep_obj=None):
         results = []
         for mode in search_strings:
             items = []
             logger.debug(_("Search Mode: {mode}".format(mode=mode)))
             for search_string in {*search_strings[mode]}:
-
                 if mode == "Season":
-                    search_string = re.sub(r"(.*)S0?", r"\1Saison ", search_string)
+                    search_string = re.sub(r"(.*)S[0-9]", r"\1Saison ", search_string)

                 if mode != "RSS":
-                    logger.debug("Search string: {0}".format(search_string))
+                    logger.debug(_("Search String: {search_string}".format(search_string=search_string)))

                     search_url = self.url + "/recherche/" + search_string
                 else:
@@ -43,17 +105,24 @@ class Provider(TorrentProvider):
                     continue

                 with BS4Parser(data, "html5lib") as html:
-                    torrent_rows = html(class_=re.compile("ligne[01]"))
+                    torrent_rows = html.find("table", {"class": "table-corps"})
+                    if torrent_table:
+                        torrent_rows = torrent_table.findAll("tr")
+                    else:
+                        torrent_rows = None
+
+                    if not torrent_rows:
+                        continue
                     for result in torrent_rows:
                         try:
                             title = result.find(class_="titre").get_text(strip=True).replace("HDTV", "HDTV x264-CPasBien")
                             title = re.sub(r" Saison", " Season", title, flags=re.I)
-                            tmp = result.find("a")["href"].split("/")[-1].replace(".html", ".torrent").strip()
-                            download_url = self.url + "/telechargement/{0}".format(tmp)
+                            tmp = result.find("a")["href"]
+                            download_url = urljoin(self.url, self._retrieve_dllink_from_url(urljoin(self.url, tmp)))
                             if not all([title, download_url]):
                                 continue

-                            seeders = try_int(result.find(class_="up").get_text(strip=True))
+                            seeders = try_int(result.find(class_="seed_ok").get_text(strip=True))
                             leechers = try_int(result.find(class_="down").get_text(strip=True))
                             if seeders < self.minseed or leechers < self.minleech:
                                 if mode != "RSS":

```

### PR DESCRIPTION
**Pull request by ghost**
Fixes #8082 #8083 #8084

Proposed changes in this pull request:
- quick fix on providers cpasbien torrent9 and torrent911

- [ ] PR is based on the SickChill:ProvidersUpdate202208 branch

### CODE REVIEW

**Comment by ghost** [comment]
Search works for those 3 providers.
I also tried to check torrent size from convert_size with a logger.debug(f"file size: {size}") but it broke sickchill.

```
2022-08-23 04:31:17 DEBUG :: WEBSERVER :: Failed doing webui request "themes/default/img/down_jpg": Traceback (most recent call last):
  File "/opt/sickchill/sickchill/views/index.py", line 159, in get
    method = getattr(self, route)
AttributeError: 'WebRoot' object has no attribute 'themes/default/img/down_jpg'
```
Those 3 providers do not use french SI units but rather legacy, convert returns -1 atm.

```python
#units = ["B", "KB", "MB", "GB", "TB", "PB"]
units = ["o", "Ko", "Mo", "Go", "To", "Po"]
size = convert_size("525.19 MB", units=units) or -1
print(size)
-1
```

For torrent9 there is a lot of redirects.
**https://www.torrent9.re** redirects to **http://ww1.torrent9.re** that redirect to **https://ww1.torrent9.re**

**Comment by miigotu** [comment]
Im going to rewrite the providers search tests, so that we can start seeing when providers are failing.

File: sickchill/oldbeard/providers/torrent9.py [hunk 1]
```diff
@@ -122,7 +122,7 @@ def search(self, search_strings, age=0, ep_obj=None):
                             if not all([title, download_url]):
                                 continue
 
-                            seeders = try_int(result.find(class_="seed_ok").get_text(strip=True))
```
Comment (miigotu) [inline sickchill/oldbeard/providers/torrent9.py:125]: class seed_ok was fine here also.
image

Comment (miigotu) [reply sickchill/oldbeard/providers/torrent9.py:125]: Nevermind, I was checking torrent9 not torrent911 xD

**Pull request merged**